src/common/lib/client/realtimechannel.ts

```diff
@@ -699,6 +700,10 @@ class RealtimeChannel extends Channel {
       state + (reason ? '; reason: ' + reason : '')
     );
 
+    if (state !== 'attaching' && state !== 'suspended') {
+      this.retryCount = 0;
```

Comment:
Is this supposed to be `if (state !== 'attached' && state !== 'suspended')` @owenpearson ?

Reply:
Hey, so this was intentional and will work correctly. In the channel retry sequence the channel state alternates between `'attaching'` and `'suspended'` so this condition is met whenever the channel leaves the retry sequence for whatever reason.

It would probably also work to have `if (state === 'attached' || state === 'failed')` (note that it's `'failed'` here rather than `'suspended'`) although I'm not 100% sure whether it's possible for the channel to transition to `'detaching'` or `'detached'` during the retry sequence so it may be necessary to include one or both of those in that condition too.

Reply:
I don't think it tries to re-attach on the channel `failed`. I have tried the same myself, when the channel doesn't have enough capability available, it will go into a `failed` state and will not re-attach.

Reply:
Yeah you're right, I'm not saying that it tried to re-attach on failed. This condition is supposed to be for when the channel *leaves* the retry sequence so that it can reset the retry counter whenever that happens. Checking that the channel state is not `attaching` or `suspended` is IMO the easiest way for us to know when the retry sequence has ended.

Reply:
I think it would work just fine for `if (state === 'attached' || state === 'failed')`

Reply:
I just checked and for ably-js if the user calls `RealtimeChannel.detach` while the client is `suspended` the channel becomes `detached, and if you call `detach` while the client is `attaching` the channel becomes `detaching` so it seems possible for the client to enter the retry sequence again without ever becoming `attached` or `failed`. It might work differently in ably-dotnet though. Also worth considering what would happen if the client receives a `DETACHED` protocol message during the retry sequence.

Reply:
Although, we can say before `detaching` state would always be `attached` which resets `retryCount` to zero.

Reply:
> Although, we can say before detaching state would always be attached which resets retryCount to zero.

This isn't necessarily true for ably-js, see my comment above

Reply:
yep .. that was previous assumption as per old code.

Reply:
> So, as of now, it can become a bug if the channel state engine follows a different pattern for the retry sequence.

IIRC the spec is pretty explicit that during retries the channel state alternates between `attaching` and `suspended` so I would be surprised if the condition I used here didn't work for the other SDKs.

Reply:
Well, we do know as of now that during suspended it can go into other states too ... It does alternate between these two states ... Though for resetting retry thats not the right condition. The right condition should be the state where it shouldn't retry. That would be `attached` and `failed`